soql-standalone-parser/src/main/scala/com/socrata/soql/parsing/RecursiveDescentParser.scala

```diff
@@ -0,0 +1,1521 @@
+                  // We are now committed to parsing a user-star-selection
+                  val ParseResult(r2, exceptions) = selectExceptions(r.rest.rest.rest, userIdentifier)
+                  result += StarSelection(Some(intoQualifier(ti)), exceptions).positionedAt(star.position)
+                  loop(r2, commaFirst = true)
+                case _ =>
+                  // Whoops, not a star selection, revert
+                  reader.addAlternates(TABLEIDSTARSET)
+                  reader
+              }
+            case _ =>
+              reader.addAlternates(TABLEIDSTARSET)
+              reader
+          }
+        case star@STAR() =>
+          val ParseResult(r2, exceptions) = selectExceptions(r.rest, userIdentifier)
+          result += StarSelection(None, exceptions).positionedAt(star.position)
+          loop(r2, commaFirst = true)
+        case _ =>
+          reader.addAlternates(TABLEIDSTARSET)
+          reader
+      }
+    }
+
+    val finalReader = loop(reader, commaFirst)
+    ParseResult(finalReader, result.result())
+  }
+
+  private def expressionSelectList(reader: Reader, commaFirst: Boolean): ParseResult[Seq[SelectedExpression]] = {
+    // similar to the user select list, we need to track our commits
+    val result = Vector.newBuilder[SelectedExpression]
+
+    @tailrec
+    def loop(reader: Reader, commaFirst: Boolean): Reader = {
+      // Unlike the user select list above, if we need to see a comma
+      // and we see one, we're committed to parsing an expression.
+      // Otherwise we might just be at the end of the list!
+
+      val r =
+        if(commaFirst) {
+          reader.first match {
+            case COMMA() =>
+              reader.rest
+            case _ =>
+              // We require a comma but do not see one - revert.
+              reader.addAlternates(COMMASET)
+              return reader
+          }
+        } else {
+          reader
+        }
+
+      // This is written in a slightly strange way to placate the
+      // tail-recursion checker (it's still lexically tail-recursive
+      // if the `catch` does a `return reader` but apparently that
+      // does something internally which causes it not to be)
+
+      val exprResult =
+        try {
+          Some(topLevelExpr(r))
+        } catch {
+          case e: ParseException if !commaFirst && e.reader == r =>
+            // ok, so if the parse failed right at the start we'll
+            // assume we're good at the end of the select list and are
+            // good to move on...
+            None
+        }
+
+      exprResult match {
+        case Some(ParseResult(r2, e)) =>
+          r2.first match {
+            case AS() =>
+              // now we must see an identifier...
+              val ParseResult(r3, identPos) = alias(r2.rest)
+              result += SelectedExpression(e, Some(identPos))
+              loop(r3, commaFirst = true)
+            case _ =>
+              r2.addAlternates(ASSET)
```

Comment:
I understand why it's named `ASSET` and I still hate it.

Reply:
Yeah, I should probably underscore all of the *SET constants...